source/common/memory/debug.h

```diff
@@ -0,0 +1,48 @@
+#pragma once
+
+#include <cstdint>
+
+namespace Envoy {
+namespace Memory {
+
+class Debug {
+public:
+  // Instantiate to force-load the memory debugging module. This is called
+  // whether or not memory-debugging is enabled, which is controlled by ifdefs
+  // in mem_debug.cc.
+  Debug();
+
+  // Returns the number of bytes used -- if memory debugging is enabled.
+  // Otherwise returns 0.
+  static uint64_t bytesUsed();
+};
+
+// Centralized ifdef logic to determine whether this compile has memory
+// debugging. This is exposed in the header file for testing.
+
+// We don't run memory debugging for optimizd builds to avoid impacting
+// production performance.
+#ifndef NDEBUG
```

Comment:
I think, I'd prefer if users had to opt-in into this via `--define debug_memory=enabled` or something like that, instead of this being automatically injected to all non-optimized / non-sanitized builds with `--define tcmalloc=disabled`.